### DISCUSSION TITLE
How can I modify a file upload in a record hook?

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1405] [Q&A]
I want to save blog posts as HTML with base64 encoded images inline.  My old system uses bluemonday to sanitize the body before saving it on the filesystem.  I'm thinking that I could use the Pocketbase built-in file uploading system for this.  So I need to be able to modify / filter the uploaded file, and to make sure that it ends up taking the place of the original one.  How would I handle this?

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand the question.

Are you trying to "embed" an uploaded multipart-file as a base64 image in some record text field?
If that's the case, then you can read the `e.HttpContext.Request().MultipartForm` inside the hook event and "extract" the files you want to embed.

**Reply by dschissler** [reply]
The article body is text (with images encoded into it as base64), but that isn't important.  I don't want to clog the database with this data, so I'm storing it in the filesystem.  I was responding by sending everything at once via the multipart form response.

So I can adapt this way to Pocketbase by just moving things around a bit and porting it from Chi to Echo.

But I thought that perhaps I could use Pocketbase's built-in uploading system to store this attached data.  The issue is that Pocketbase doesn't provide access to this uploaded file within the record event hooks.  So there is no way to modify it before it's stored unless I go through the HTTP request hooks.

The method that I found which might be able to accomplish this by this:

```javascript
fh := e.HttpContext.Request().MultipartForm.File["fileField"][0]
```

Then I would somehow access this file and then to overwrite it / filter it to be saved after bluemonday modifies the data.  Is that possible?  I'm not familiar with intercepting file uploads in this way.  Is it possible to replace the uploaded file in the request?

Also a related matter, is there any built-in way that Pocketbase can respond with record data with its file data in a multipart response, or do I need to create that custom API myself?

**Reply by dschissler** [reply]
Actually, I was trying to keep it simple (on my end) with the base64 encoded images on my old system because it makes it super easy in the Javascript end, but perhaps it would actually easier to use the pocketbase feature that allows multiple images per record.  That was much more involved than what I was willing to do for that system when I was just learning the ecosystem and language.

That is something that I can pursue.  This is a private non-job project that I'm working on when I have a few moments from my other professional Pocketbase project.  So it's not urgent.

I think thought that people are going to need a pattern, and documentation for how to intercept files to rewrite them, or to resize them.  For example, 1) a system that resizes images uploads to a maximum size, 2) filtering huge text uploads, 3) processing PDF's by breaking it into 1-page files, 4) Many situations in which filtering the upload is necessary.

**Reply by ganigeorgiev** [reply]
Currently the only way to modify/read the uploaded files is in the `*Request` hooks using the `Request().MultipartForm` instance. I'll consider adding the resolved `RecordUpsert` uploaded files to the `OnRecordBeforeCreateRequest` and `OnRecordBeforeUpdateRequest` hooks sometime in the future to make it a little bit easier.

**Reply by dschissler** [reply]
I think that a simple example for resizing an image to a maximum storage size would give people a good idea of how to intercept.  The image conversion example could be basically copy pasted from some library to make it no larger than 1000x1000.  Then perhaps it's as simple of a matter of replacing a file in the slice.

Probably giving examples for both hooks could go a long way.  It could be that the `OnRecordBeforeCreateRequest` event is cleaner to read as so is the first choice, but that there might be other exposed data that requires Request data.

**Comment by dschissler** [comment]
@ganigeorgiev Hey I have an idea for the direction that would allow me to continue doing it the way that it was.  But I need to know if the pocketbase API will allow fields to be set that will show up in the record hook which don't correspond to any database field.  For example, if `body` was the article body that had base64 encoded images.  If this was available in the record hook then I could use it to save this `body` content in a file outside of the normal Pocketbase `storage` directory.  Then in the api request event I could load the article body file into the `body` field.

Do these extra values find their way to the record event?  Or is it only populated with existing database values from the request?

I could experiment with this to find out, but I'm still not sure how that I would like to do it.

**Reply by ganigeorgiev** [reply]
The `models.Record` struct when loaded from the request is populated only with the collection schema defined fields (+ some special system fields like `id`, `email`, etc.).

You can get the raw parsed request body from the `e.HttpContext`. There is also a helper:

```go
r := apis.RequestData(e.HttpContext)
log.Println(r.Data["body"])
```

**Reply by dschissler** [reply]
If I were to set a record field in this Request Before event then would it be available in the model event?

**Reply by ganigeorgiev** [reply]
Yes. If you manually set it, then you can can access it as usual, for example `record.GetString("body")`.
